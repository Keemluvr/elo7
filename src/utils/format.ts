/**
 * Capitalizes the first letter of the input string and converts the rest to lowercase.
 *
 * @param {string} string - The input string to be capitalized.
 * @return {string} The capitalized string.
 */
export const capitalize = (string: string) =>
  string.charAt(0).toUpperCase() + string.slice(1).toLowerCase();

/**
 * Capitalizes each word in a string, including words separated by hyphens.
 *
 * @param {string} string - The input string to be capitalized.
 * @return {string} The capitalized string with each word capitalized and words separated by hyphens preserved.
 */
export const capitalizeWords = (string: string) => {
  return string
    .split(" ")
    .map((word) => {
      if (word.includes("-")) {
        const parts = word.split("-");
        return parts.map((part) => capitalize(part)).join("-");
      }
      return capitalize(word);
    })
    .join(" ");
};

/**
 * Formats a location string into a city, state, and country.
 *
 * @param {string} location - The location string to format.
 * @return {string} The formatted location string.
 */
export const formatLocation = (location: string) => {
  const [city, state, country] = location.split(", ");
  if (!city || !country) return "";
  return `${capitalizeWords(city)}, ${capitalizeWords(country)}`;
};

/**
 * Replaces specific substrings in a given string with their corresponding uppercase versions.
 *
 * @param {string} string - The input string to be processed.
 * @return {string} The modified string with the special cases replaced.
 */
export const specialCases = (string: string) => {
  return string
    .replace(/fp&a/gi, "FP&A")
    .replace(/Android/gi, "Android")
    .replace(/Ios/gi, "iOS")
    .replace(/devops/gi, "DevOps")
    .replace(/Sre/gi, "SRE");
};
