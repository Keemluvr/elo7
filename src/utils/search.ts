interface Data<T> {
  [key: string]: T[];
}

/**
 * Searches the data for entries that match all the terms in the query.
 *
 * @param {Data<T>} data - The data to search through.
 * @param {string | null} query - The search query.
 * @param {(keyof T)[]} properties - The properties to search within.
 * @return {Data<T> | undefined} The filtered data based on the search query.
 */
export const searchByCombinedTerms = <T>(
  data: Data<T>,
  query: string | null,
  properties: (keyof T)[]
): Data<T> | undefined => {
  if (!query) return;

  const terms = query.toLowerCase().split(" ");
  const filteredData: Data<T> = {};

  for (const key in data) {
    const filteredEntries = data[key].filter((entry: T) => {
      const entryString = properties
        .map((prop) => (entry[prop] as string).toLowerCase())
        .join(" ");
      return terms.every((term) => entryString.includes(term));
    });

    if (filteredEntries.length > 0) {
      filteredData[key] = filteredEntries;
    }
  }
  return Object.keys(filteredData).length > 0 ? filteredData : undefined;
};
