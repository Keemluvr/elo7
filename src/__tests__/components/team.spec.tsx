import { render } from "@testing-library/react";
import { verifyAttributes } from "../utils/verifyAttributes";
import Team from "@/components/team";

describe("Team component", () => {
  describe("Section", () => {
    const sectionRoleName = "seção da apresentação do time";

    test("should render the section component", () => {
      const { getByRole } = render(<Team />);
      const sectionElement = getByRole("region", { name: sectionRoleName });
      expect(sectionElement).toBeInTheDocument();
    });

    test("should render the section with correct tag name", () => {
      const { getByRole } = render(<Team />);
      const sectionElement = getByRole("region", { name: sectionRoleName });
      expect(sectionElement.tagName).toBe("SECTION");
    });
  });

  describe("Title", () => {
    const titleText = "Conheça nosso time fora de série";

    test("should render the title", () => {
      const { getByText } = render(<Team />);
      const titleElement = getByText(titleText);
      expect(titleElement).toBeInTheDocument();
    });

    test("should render the title with correct tag name", () => {
      const { getByText } = render(<Team />);
      const titleElement = getByText(titleText);
      expect(titleElement.tagName).toBe("H2");
    });

    test("should render the title with correct properties", () => {
      const { getByText } = render(<Team />);
      const titleElement = getByText(titleText);
      expect(titleElement).toHaveClass("team-title");
    });
  });

  describe("Cards", () => {
    const teamMembers = [
      {
        name: "ana",
        alt: "Ana, uma mulher sorrindo enquanto trabalha em um ambiente colaborativo.",
      },
      {
        name: "joao",
        alt: "João, um homem de barba ruiva e camisa branca, sorrindo com os braços cruzados em um ambiente de escritório, com duas pessoas ao fundo conversando.",
      },
      {
        name: "maria",
        alt: "Maria, uma mulher sorrindo enquanto segura um laptop em um ambiente de escritório moderno.",
      },
      {
        name: "pedro",
        alt: "Pedro, um homem com cabelo cacheado e barba, usando fones de ouvido e uma camisa azul, sentado em frente a um laptop e sorrindo.",
      },
    ];

    test.each(teamMembers)("should render the $name's photo", ({ alt }) => {
      const { getByAltText } = render(<Team />);
      const teamMemberElement = getByAltText(alt);
      expect(teamMemberElement).toBeInTheDocument();
    });

    test.each(teamMembers)(
      "should render the correct attributes for the $name's photo",
      ({ name, alt }) => {
        const { getByAltText } = render(<Team />);
        const teamMemberElement = getByAltText(alt);
        const attributes = [
          ["aria-label", `imagem membro da equipe: ${name}`],
          ["loading", "lazy"],
        ];
        verifyAttributes(teamMemberElement, attributes);
        expect(teamMemberElement).toHaveAttribute("srcset");
        expect(teamMemberElement.getAttribute("srcset")).not.toBe("");
      }
    );
  });

  test("matches snapshot", async () => {
    const { asFragment } = render(<Team />);
    expect(asFragment()).toMatchSnapshot();
  });
});
