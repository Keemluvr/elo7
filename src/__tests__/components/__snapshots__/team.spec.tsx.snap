// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Team component matches snapshot 1`] = `
<DocumentFragment>
  <section
    aria-label="seção da apresentação do time"
  >
    <div
      class="grid-container"
    >
      <h2
        class="grid-item-full-width team-title"
      >
        Conheça nosso time fora de série
      </h2>
      <div
        class="grid-item-full-width team-content"
      >
        <div
          class="team-member-card"
        >
          <img
            alt="Ana, uma mulher sorrindo enquanto trabalha em um ambiente colaborativo."
            aria-label="imagem membro da equipe: ana"
            class="team-member-card-image"
            data-nimg="fill"
            decoding="async"
            loading="lazy"
            sizes="100vw"
            src="/_next/image?url=%2Fimages%2Fteam%2Fana.webp&w=3840&q=100"
            srcset="/_next/image?url=%2Fimages%2Fteam%2Fana.webp&w=640&q=100 640w, /_next/image?url=%2Fimages%2Fteam%2Fana.webp&w=750&q=100 750w, /_next/image?url=%2Fimages%2Fteam%2Fana.webp&w=828&q=100 828w, /_next/image?url=%2Fimages%2Fteam%2Fana.webp&w=1080&q=100 1080w, /_next/image?url=%2Fimages%2Fteam%2Fana.webp&w=1200&q=100 1200w, /_next/image?url=%2Fimages%2Fteam%2Fana.webp&w=1920&q=100 1920w, /_next/image?url=%2Fimages%2Fteam%2Fana.webp&w=2048&q=100 2048w, /_next/image?url=%2Fimages%2Fteam%2Fana.webp&w=3840&q=100 3840w"
            style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
          />
        </div>
        <div
          class="team-member-card"
        >
          <img
            alt="João, um homem de barba ruiva e camisa branca, sorrindo com os braços cruzados em um ambiente de escritório, com duas pessoas ao fundo conversando."
            aria-label="imagem membro da equipe: joao"
            class="team-member-card-image"
            data-nimg="fill"
            decoding="async"
            loading="lazy"
            sizes="100vw"
            src="/_next/image?url=%2Fimages%2Fteam%2Fjoao.webp&w=3840&q=100"
            srcset="/_next/image?url=%2Fimages%2Fteam%2Fjoao.webp&w=640&q=100 640w, /_next/image?url=%2Fimages%2Fteam%2Fjoao.webp&w=750&q=100 750w, /_next/image?url=%2Fimages%2Fteam%2Fjoao.webp&w=828&q=100 828w, /_next/image?url=%2Fimages%2Fteam%2Fjoao.webp&w=1080&q=100 1080w, /_next/image?url=%2Fimages%2Fteam%2Fjoao.webp&w=1200&q=100 1200w, /_next/image?url=%2Fimages%2Fteam%2Fjoao.webp&w=1920&q=100 1920w, /_next/image?url=%2Fimages%2Fteam%2Fjoao.webp&w=2048&q=100 2048w, /_next/image?url=%2Fimages%2Fteam%2Fjoao.webp&w=3840&q=100 3840w"
            style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
          />
        </div>
        <div
          class="team-member-card"
        >
          <img
            alt="Maria, uma mulher sorrindo enquanto segura um laptop em um ambiente de escritório moderno."
            aria-label="imagem membro da equipe: maria"
            class="team-member-card-image"
            data-nimg="fill"
            decoding="async"
            loading="lazy"
            sizes="100vw"
            src="/_next/image?url=%2Fimages%2Fteam%2Fmaria.webp&w=3840&q=100"
            srcset="/_next/image?url=%2Fimages%2Fteam%2Fmaria.webp&w=640&q=100 640w, /_next/image?url=%2Fimages%2Fteam%2Fmaria.webp&w=750&q=100 750w, /_next/image?url=%2Fimages%2Fteam%2Fmaria.webp&w=828&q=100 828w, /_next/image?url=%2Fimages%2Fteam%2Fmaria.webp&w=1080&q=100 1080w, /_next/image?url=%2Fimages%2Fteam%2Fmaria.webp&w=1200&q=100 1200w, /_next/image?url=%2Fimages%2Fteam%2Fmaria.webp&w=1920&q=100 1920w, /_next/image?url=%2Fimages%2Fteam%2Fmaria.webp&w=2048&q=100 2048w, /_next/image?url=%2Fimages%2Fteam%2Fmaria.webp&w=3840&q=100 3840w"
            style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
          />
        </div>
        <div
          class="team-member-card"
        >
          <img
            alt="Pedro, um homem com cabelo cacheado e barba, usando fones de ouvido e uma camisa azul, sentado em frente a um laptop e sorrindo."
            aria-label="imagem membro da equipe: pedro"
            class="team-member-card-image"
            data-nimg="fill"
            decoding="async"
            loading="lazy"
            sizes="100vw"
            src="/_next/image?url=%2Fimages%2Fteam%2Fpedro.webp&w=3840&q=100"
            srcset="/_next/image?url=%2Fimages%2Fteam%2Fpedro.webp&w=640&q=100 640w, /_next/image?url=%2Fimages%2Fteam%2Fpedro.webp&w=750&q=100 750w, /_next/image?url=%2Fimages%2Fteam%2Fpedro.webp&w=828&q=100 828w, /_next/image?url=%2Fimages%2Fteam%2Fpedro.webp&w=1080&q=100 1080w, /_next/image?url=%2Fimages%2Fteam%2Fpedro.webp&w=1200&q=100 1200w, /_next/image?url=%2Fimages%2Fteam%2Fpedro.webp&w=1920&q=100 1920w, /_next/image?url=%2Fimages%2Fteam%2Fpedro.webp&w=2048&q=100 2048w, /_next/image?url=%2Fimages%2Fteam%2Fpedro.webp&w=3840&q=100 3840w"
            style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
          />
        </div>
      </div>
    </div>
  </section>
</DocumentFragment>
`;
